### ISSUE TITLE
__callee__ behaves differently from MRI

### ISSUE THREAD

**Issue by guai** [jruby/jruby#2305]
jruby 1.7.13 and 1.7.16.1 both in --2.0 mode

``` ruby
class Foo
    def foo
        {__method__ => __callee__}
    end
    alias bar foo
end
raise 'fail' if Foo.new.foo == Foo.new.bar
```

They must not be equal.

**Comment by subbuss** [comment]
This seems to be the reason for one of failures seen in builds (https://travis-ci.org/jruby/jruby/jobs/50726061) that I investigated .. 

See lib/ruby/stdlib/cgi/html.rb:19

```
    def nn_element_def(attributes = {}, &block)
      nn_element(__callee__, attributes, &block)
    end
```

The tests don't fail in interpreted mode, but in compiled mode, it fails intermittently depending on whether the JIT gets to it or not.

But, FWIW, this has been failing for a while, but not sure why it is triggering failed builds now.

**Comment by personnel** [comment]
Same problem.Metaprogramming

**Comment by headius** [comment]
There might be some way to support this in IR (if we can pass the aliased name through into all bodies as well as the actual method name) but this is a really troublesome feature to support. We don't save the actual called name, and I'm not sure I agree that we should.

@guai Your example, if it worked on JRuby, would be almost unusably slow. `__method__` and `__callee__` are implemented using `caller` logic, which is very expensive. Neither of them are particularly fast on MRI either, but they're orders of magnitude slower than that on JRuby.

**Comment by headius** [comment]
One possible way to implement this: when a method has been aliased, generate a new jitted body that has both the alias name and the original name in its signature. We can mine that out of caller backtrace then. It still will be really slow due to the need to generate a backtrace.

**Comment by headius** [comment]
Another possible way to implement this: pass the original method (AliasMethod in this case) through into the wrapped method so it always knows what method it was called as. It can report __callee__ properly then without too much tricker. This approach will be a bit more complicated since we'd need the method to be available for evals and blocks inside a given method too.

**Comment by headius** [comment]
cc @subbuss @enebo for more ideas.

**Comment by enebo** [comment]
@headius I had worked on trying to pass the callsite through the call locally on a branch in one effort to make profiler easier.  This would then be a trivial feature.  Perhaps this is one more reason to pass callsite through call paths.

**Comment by headius** [comment]
@enebo That would indeed solve this case. And since CallSite is a pretty good place to store static call info, we might be able to use it for other purposes too.

**Comment by headius** [comment]
This is still an outstanding issue with no easy fix. The problem is largely that when we have an alias, we do not pass the called name along at all:

https://github.com/jruby/jruby/blob/3707e81f9833ee9b5d07b38edfc781502bdc418f/core/src/main/java/org/jruby/internal/runtime/methods/AliasMethod.java#L78

Because of this, it doesn't even make it as far as the interpreter/bytecode or into the frame.

Fixing this would at least require we pass through the called name along with the "super" name (the actual name of the target method) or else acquire the super name some other way that allows us to pass the called name through.

**Comment by headius** [comment]
I have pushed a short-term fix in #7702 that encodes both the aliased name and original name in the name passed through the call stack. Anywhere that consumes either the "super name" or the "callee name" must decode the composite name in oder to get those others out. It appears to be working properly, but will likely need more bake time before release.
